# Neighbourhood segmentation and clustering-Toronto

In [30]:
import pandas as pd

## Question 1

## Importing data

In [31]:
import requests

In [32]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [33]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [34]:
len(wiki_data), type(wiki_data)

(3, list)

Requirement of first table alone,so dropping other tables

In [35]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [36]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [37]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [38]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [39]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [40]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [41]:
df.shape

(103, 3)

Answer to Q1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [42]:
pip install geocoder

     |████████████████████████████████| 98 kB 7.7 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [43]:
import geocoder # import geocoder

Importing the CSV file from the URL

In [44]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [45]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes to join

In [46]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [47]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [48]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [49]:
combined_data.shape

(103, 5)

We get 103 rows of data from inner join

# Question 3

Exploring and clustering Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [50]:
from geopy.geocoders import Nominatim

In [51]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Visualizing the map of Toronto

In [57]:
!pip install folium
import folium

print('Folium installed and imported!')

     |████████████████████████████████| 93 kB 3.2 MB/s eta 0:00:011
Folium installed and imported!


In [58]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto


Initializing Foursquare API credentials

In [59]:
CLIENT_ID = '5TOQ1V2HVGFAHQVVBQ20P04AMSZ3VEK2WT0N5QRKOH1QNBHC' 
CLIENT_SECRET = 'UGUBNVKHM3VOMWTZ114H33DHQQGE0WPLOE5XFB3GEWNGI1IJ'
VERSION = '20201028' # Foursquare API version

Create a function to get all the venue categories in Toronto

In [60]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [61]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [62]:
venues_in_toronto.shape

(1317, 5)

In [63]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

Checking the Venues based on Neighbourhood

In [64]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1301  Mimico NW, The Queensway West, South of Bloor,...   
1302  Mimico NW, The Queensway West, South of Bloor,...   
1303  Mimico NW, The Queensway West, South of Bloor,...   
1304  Mimico NW, The Queensway West, South of Bloor,...   
1305  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1301               43.628841               -79.520999              Wingporium   
1302               43.628841               -79.520999        South St. Burger   
1303               43.628841               -79.520999               Dollarama   
1304               43.628841               -79.520999          Healthy Planet   
1305               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1301            Wings Joint  
1302           Burger Joint  
1303         Discount Store  
1304        Supplement Shop  
1305                 Bakery  

[412 rows x 5 columns]

In [65]:
venues_in_toronto.groupby('Venue Category').max() #max venue category

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                              Studio District   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.659526               -79.340923   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.718518               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[237 rows x 4 columns]

# One Hot encoding the venue Categories

In [66]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1312                  0        0                   0             0   
1313                  0        0                   0             0   
1314                  0        0                   0             0   
1315                  0        0                   0             0   
1316                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1312               0                0                 0                    0   
1313               0                0                 0                    0   
1314               0                0                 0                    0   
1315               0                0                 0                    0   
1316               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  Turkish Restaurant  \
0                0         0  ...              0                   0   
1                0         0  ...              0                   0   
2                0         0  ...              0                   0   
3                0         0  ...              0                   0   
4                0         0  ...              0                   0   
...            ...       ...  ...            ...                 ...   
1312             0         0  ...              0                   0   
1313             0         0  ...              0                   0   
1314             0         0  ...              0                   0   
1315             0         0  ...              0                   0   
1316             0         0  ...              0                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1312                              0                 0                      0   
1313                              0                 0                      0   
1314                              0                 0                      0   
1315                              0                 0                      0   
1316                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4  

In [67]:
#Adding the neighbourhood
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Turkish Restaurant  \
0                    0             0  ...              0                   0   
1                    0             0  ...              0                   0   
2                    0             0  ...              0                   0   
3                    0             0  ...              0                   0   
4                    0             0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 238 columns]

In [68]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.045455           0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 238 columns]

Make a function to get the top most common venue categories

In [69]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [70]:
import numpy as np


There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods



In [71]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1             Sandwich Place           Coffee Shop                   Pub   
2                       Bank              Pharmacy    Chinese Restaurant   
3         Chinese Restaurant   Japanese Restaurant                  Café   
4         Italian Restaurant        Sandwich Place      Sushi Restaurant   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0        Clothing Store          Dessert Shop      Electronics Store   
1              Pharmacy                   Gym  General Entertainment   
2         Shopping Mall                 Diner         Sandwich Place   
3           Yoga Studio          Dessert Shop      Electronics Store   
4      Greek Restaurant     Indian Restaurant              Juice Bar   

         8th Most Common Venue 9th Most Common Venue   10th Most Common Venue  
0  Eastern European Restaurant   Dumpling Restaurant                Drugstore  
1                    Drugstore               Dog Run      Distribution Center  
2                Deli / Bodega            Restaurant              Supermarket  
3  Eastern European Restaurant   Dumpling Restaurant                Drugstore  
4                 Liquor Store             Locksmith  Comfort Food Restaurant

Model to cluster the neighbourhoods

In [72]:
# import k-means from clustering stage
from sklearn.cluster import KMeans
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [73]:
kmeans.labels_[0:100] #labelling of the model

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 4, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1,
       0, 1, 1, 1, 1, 1, 0, 2], dtype=int32)

Add the clustering Label column to the top 10 common venue categorie

In [74]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [75]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio      Department Store   
1          Intersection           Coffee Shop          Hockey Arena   
2                  Park                Bakery        Breakfast Spot   
3     Accessories Store              Boutique             Gift Shop   
4           Yoga Studio                  Café                   Bar   

    5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0       Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
1   Portuguese Restaurant                        Diner      Department Store   
2                     Pub               Chocolate Shop   Distribution Center   
3  Furniture / Home Store                  Event Space           Coffee Shop   
4      Italian Restaurant                     Beer Bar        Sandwich Place   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0             Drugstore             Donut Shop                Dog Run  
1          Dessert Shop     Dim Sum Restaurant    Distribution Center  
2                   Spa           Dessert Shop         Farmers Market  
3         Women's Store  Vietnamese Restaurant       Airport Terminal  
4   Distribution Center                  Diner     Mexican Restaurant

In [76]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels']) #dropping NaN values

Plotting the clusters on the map

In [78]:
import matplotlib.cm as cm
import matplotlib.colors as colors

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [79]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
49        North York             0.0                  Park   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop           Yoga Studio      Department Store   
21               Women's Store                  Pool           Yoga Studio   
35           Convenience Store          Intersection           Yoga Studio   
49  Construction & Landscaping                Bakery           Yoga Studio   
61                 Swim School              Bus Line           Yoga Studio   
64                 Yoga Studio         Deli / Bodega     Electronics Store   
66           Convenience Store           Yoga Studio      Department Store   
85                        Park                Bakery           Yoga Studio   
91                       Trail            Playground           Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21                Deli / Bodega  Eastern European Restaurant   
35             Department Store            Electronics Store   
49             Department Store            Electronics Store   
61             Department Store  Eastern European Restaurant   
64  Eastern European Restaurant          Dumpling Restaurant   
66            Electronics Store  Eastern European Restaurant   
85                Deli / Bodega  Eastern European Restaurant   
91             Department Store                 Dessert Shop   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Dumpling Restaurant             Drugstore            Donut Shop   
21          Dumpling Restaurant             Drugstore            Donut Shop   
35  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
49  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
61          Dumpling Restaurant             Drugstore            Donut Shop   
64                    Drugstore            Donut Shop               Dog Run   
66          Dumpling Restaurant             Drugstore            Donut Shop   
85          Dumpling Restaurant             Drugstore            Donut Shop   
91           Dim Sum Restaurant                 Diner        Discount Store   

   10th Most Common Venue  
0                 Dog Run  
21                Dog Run  
35             Donut Shop  
49             Donut Shop  
61                Dog Run  
64    Distribution Center  
66                Dog Run  
85                Dog Run  
91                Dog Run

cluster 2

In [80]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1          North York             1.0                 Pizza Place   
2    Downtown Toronto             1.0                 Coffee Shop   
3          North York             1.0              Clothing Store   
4    Downtown Toronto             1.0                 Coffee Shop   
7          North York             1.0                         Gym   
..                ...             ...                         ...   
98          Etobicoke             1.0                        Pool   
99   Downtown Toronto             1.0                 Coffee Shop   
100      East Toronto             1.0          Light Rail Station   
101         Etobicoke             1.0  Construction & Landscaping   
102         Etobicoke             1.0               Grocery Store   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1            Intersection           Coffee Shop          Hockey Arena   
2                    Park                Bakery        Breakfast Spot   
3       Accessories Store              Boutique             Gift Shop   
4             Yoga Studio                  Café                   Bar   
7     Japanese Restaurant           Coffee Shop            Beer Store   
..                    ...                   ...                   ...   
98                  River           Yoga Studio          Dance Studio   
99                Gay Bar            Restaurant     Indian Restaurant   
100         Auto Workshop            Comic Shop           Pizza Place   
101        Baseball Field           Yoga Studio          Dessert Shop   
102     Convenience Store                Bakery         Tanning Salon   

           5th Most Common Venue 6th Most Common Venue  \
1          Portuguese Restaurant                 Diner   
2                            Pub        Chocolate Shop   
3         Furniture / Home Store           Event Space   
4             Italian Restaurant              Beer Bar   
7                 Clothing Store   Sporting Goods Shop   
..                           ...                   ...   
98   Eastern European Restaurant   Dumpling Restaurant   
99                      Beer Bar  Ethiopian Restaurant   
100             Recording Studio            Restaurant   
101                  Escape Room     Electronics Store   
102               Discount Store         Burrito Place   

           7th Most Common Venue 8th Most Common Venue  \
1               Department Store          Dessert Shop   
2            Distribution Center                   Spa   
3                    Coffee Shop         Women's Store   
4                 Sandwich Place   Distribution Center   
7                           Café    Italian Restaurant   
..                           ...                   ...   
98                     Drugstore            Donut Shop   
99                   Escape Room             Juice Bar   
100                      Butcher         Burrito Place   
101  Eastern European Restaurant   Dumpling Restaurant   
102         Fast Food Restaurant          Burger Joint   

      9th Most Common Venue 10th Most Common Venue  
1        Dim Sum Restaurant    Distribution Center  
2              Dessert Shop         Farmers Market  
3     Vietnamese Restaurant       Airport Terminal  
4                     Diner     Mexican Restaurant  
7        Dim Sum Restaurant   Caribbean Restaurant  
..                      ...                    ...  
98                  Dog Run    Distribution Center  
99           Breakfast Spot        Bubble Tea Shop  
100              Skate Park                Brewery  
101               Drugstore             Donut Shop  
102  Thrift / Vintage Store         Hardware Store  

[88 rows x 12 columns]

cluster 3

In [81]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             2.0   Martial Arts School           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
45      Department Store     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
45               Dog Run    Distribution Center

cluster 4

In [82]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             3.0            Print Shop           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

cluster 5

In [83]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             4.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6   Dumpling Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run    Distribution Center

We have successfully cluster Toronto neighbourhood based on venue categories